core/services/log/registrations.go

```diff
@@ -0,0 +1,283 @@
+//
+// 3. The logs are attempted to be sent after every new head arrival:
+// 		Each stored log is then checked against every matched listener and is sent unless:
+//    A) is too young for that listener
+//    B) the corresponding block height is known to be already processed for that listener
+// In the normal case, each log will be only processed once, and then its corresponding head will be remembered, so it's not double-sent
+//
+// After processing the whole batch of logs considered for sending, the per-listener metadata is updated in applyListenerInfoUpdates.
+// If a re-org happens, the stored lowestAllowedBlockNumber (per-listener) is re-set,
+// so the logs from that chain are then considered unprocessed, and will be sent again.
+//
+type (
+	registrations struct {
+		registrations map[common.Address]map[common.Hash]map[Listener]*listenerMetadata // contractAddress => logTopic => Listener
+		decoders      map[common.Address]AbigenContract
+
+		// highest 'num_confirmations' per all listeners, used to know when to delete old logs
+		highestNumConfirmations uint64
+	}
+
+	// The Listener responds to log events through HandleLog, and contains setup/tear-down
+	// callbacks in the On* functions.
+	Listener interface {
+		HandleLog(b Broadcast)
+		JobID() models.JobID
+		JobIDV2() int32
+		IsV2Job() bool
+	}
+
+	// metadata structure maintained per listener, used to avoid double-sends of logs
+	listenerMetadata struct {
+		opts                     ListenerOpts
+		lowestAllowedBlockNumber uint64
+		lastSeenChain            *models.Head
+	}
+
+	// an update to listener metadata structure
+	listenerMetadataUpdate struct {
+		toUpdate                    *listenerMetadata
+		newLowestAllowedBlockNumber uint64
+	}
+)
+
+func newRegistrations() *registrations {
+	return &registrations{
+		registrations: make(map[common.Address]map[common.Hash]map[Listener]*listenerMetadata),
+		decoders:      make(map[common.Address]AbigenContract),
+	}
+}
+
+func (r *registrations) addSubscriber(reg registration) (needsResubscribe bool) {
+	addr := reg.opts.Contract.Address()
+	r.decoders[addr] = reg.opts.Contract
+
+	if reg.opts.NumConfirmations <= 0 {
+		reg.opts.NumConfirmations = 1
+	}
+
+	if _, exists := r.registrations[addr]; !exists {
+		r.registrations[addr] = make(map[common.Hash]map[Listener]*listenerMetadata)
+	}
+
+	topics := make([]common.Hash, len(reg.opts.Logs))
+	for i, log := range reg.opts.Logs {
+		topic := log.Topic()
+		topics[i] = topic
+
+		if _, exists := r.registrations[addr][topic]; !exists {
+			r.registrations[addr][topic] = make(map[Listener]*listenerMetadata)
+			needsResubscribe = true
+		}
+		r.registrations[addr][topic][reg.listener] = &listenerMetadata{
+			opts:                     reg.opts,
+			lowestAllowedBlockNumber: uint64(0),
+		}
+	}
+
+	r.maybeIncreaseHighestNumConfirmations(reg.opts.NumConfirmations)
+	return
+}
+
+func (r *registrations) removeSubscriber(reg registration) (needsResubscribe bool) {
```

Comment:
This does not look thread-safe. Perhaps it doesn't matter? If not can we add a comment explaining why

Reply:
it is thread-safe - it's being called only from the main event loop, on receiving a message from a mailbox (onRmSubscribers)